### ISSUE TITLE
[Q] Text Editing / VIM Commands

### ISSUE THREAD

**Issue by kghenderson** [pbek/QOwnNotes#1633]
#### General question

Hi, 

First off, thanks for the amazing software, I really enjoy it!

I'm hoping to figure out how to add some more basic editor functions though, even simple things like move line up/down for outlining within a note.

I started down the path of using scripting for this and made a bit of progress, but I thought I'd reach out to see if this was really the best strategy. I noticed that there's a VIM mode within the editor and thought there might be a way to access these functions maybe?

Thanks!
Kevin

**Comment by pbek** [comment]
Thank you for your kind words.

Can you please be a bit more specific of what feature you would like, how they should work (maybe with an example where you saw them) and possible keyboard shortcuts for them.

> I noticed that there's a VIM mode within the editor and thought there might be a way to access these functions maybe?

which functions and access how?

**Comment by kghenderson** [comment]
I do a lot with simple, nested bulleted lists in Markdown, I'd like to move the selected line up and down and indent/outdent it left and right using keyboard shortcuts. 

e.g. 

- A 
  - C
  - B  <-- be on this line, press ctrl+shift+up and have it raise up a line
- D

Writing this with a JavaScript plugin, I was thinking of splitting the whole file into an array (using the the sort-lines script as a model), then swapping line positions, but couldn't figure out how to get the current line/selected position index.

I don't really know much about QON's Vim Mode, but I've used Vim and it has many features for moving around and selecting text and thought maybe they could be accessed somehow. 

Regarding shortcuts, I think Ctrl+Shift+Arrow Keys are the most common in other editors like Sublime Text, but personally like to customize them to [Ctrl+Shift+I] for Up, [Ctrl+Shift+Comma] for Down, [Ctrl+Shift+J] for Left/Outdent, and [Ctrl+Shift+L] for Right/Indent so I don't need to leave homerow - but I don't mind setting custom shortcuts :-)

**Comment by pbek** [comment]
Maybe you can get @Waqar144 interested. :grin: 

> couldn't figure out how to get the current line/selected position index.

There is https://docs.qownnotes.org/en/develop/scripting/README.html#get-the-start-position-of-the-current-selection-in-the-note-text-edit

**Comment by Waqar144** [comment]
> I do a lot with simple, nested bulleted lists in Markdown, I'd like to move the selected line up and down

This is already implemented, Ctrl + Shift + Up/Down

> and indent/outdent it left and right using keyboard shortcuts.

This is also in place for some things. I think the shortcut for it is: Shift + Tab (not sure)

> Writing this with a JavaScript plugin, I was thinking of splitting the whole file into an array (using the the sort-lines script as a model), then swapping line positions, but couldn't figure out how to get the current line/selected position index.

All the text manipulation / iteration is done using a `QTextcursor`. We have `textCursor` exposed to the QML side. Current line can be obtained using `textCursor.block()` and current position can be obtained using `textCursor.position()`

> I don't really know much about QON's Vim Mode, but I've used Vim and it has many features for moving around and selecting text and thought maybe they could be accessed somehow.

We use FakeVim for vim emulation, and I haven't looked at its code but I am pretty sure it manipulates stuff using the same methods. 
@pbek can Fakevim api be exposed to scripting? I am not so sure if it's a great idea though. And it might not be easy since it's already working through a proxyWidget

**Comment by pbek** [comment]
@kghenderson, https://www.qownnotes.org/shortcuts/QOwnNotes

> @pbek can Fakevim api be exposed to scripting? I am not so sure if it's a great idea though. And it might not be easy since it's already working through a proxyWidget

I don't know, I haven't played around with it a lot beside integrating it...

**Comment by kghenderson** [comment]
Thanks so much for responding guys, I really appreciate it.
I'm sorry for the naive questions, but is there perhaps an example somewhere of accessing the QTextCursor object then from within a script? 

Let's say for example I wanted to write a task creation script where Ctrl+T (or something) added four spaces added "    - [ ] sample" to the beginning of the current line (not the selection). 

I know this is really basic, just trying to get my bearings. Thank you.

**Comment by pbek** [comment]
By using https://docs.qownnotes.org/en/develop/scripting/README.html#select-the-current-line-in-the-note-text-edit you could select the current line and then you could add your text at the beginning of that selection and write the new text into the note text edit.

**Comment by Waqar144** [comment]
https://github.com/qownnotes/scripts is where all publicly shared QON scripts live. There isn't a single script that's using textCursor though :thinking:

**Comment by pbek** [comment]
Is textCursor documented? Haven't found it. Is that even exposed in the api? 😁

**Comment by Waqar144** [comment]
> Is textCursor documented? Haven't found it. Is that even exposed in the api? 😁

Yes, I can't see that in the api lol. But there are indirect methods for applying modifications to text. 

I think I read somewhere in the issues about textCursor not accessible despite being exposed to QML

**Comment by pbek** [comment]
Were is it exposed?

**Comment by Waqar144** [comment]
I think using document(), but it's not available now.
